### ISSUE TITLE
sql: subquery_retry test occasionally fails with one duplicate

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#4607]
As part of the #4472 I wanted to enable the `subquery_retry` parallel test. It usually passes but I have seen occasional failures. Every time it failed, the failure showed exactly one duplicate value:

```
--- FAIL: TestParallel (48.88s)
	logic_test.go:654: partestdata/subquery_retry/final:2: expected:
		"0"  
		but found:
		"1"  
```
This might be the same root cause as #4560 but I am filing a separate issue until we know for sure.

I tried using `make stress` to run the test in a loop and found no failures.

I tried making the test longer by increasing the repeat count in https://github.com/cockroachdb/cockroach/blob/master/sql/partestdata/subquery_retry/txn

With `repeat 500`, I hit the failure 2 times and then did not hit it again after a few more runs. I increased to `repeat 2000` (and set `TESTTIMEOUT=30m`) and it still didn't fail.

To run, the `run txn txn txn` line must be uncommented in https://github.com/cockroachdb/cockroach/blob/master/sql/partestdata/subquery_retry/main

**Comment by tschottdorf** [comment]
`make stress PKG=./sql TESTS=Parallel TESTFLAGS='-partestdata sql/partestdata/subquery_retry' STRESSFLAGS='-maxfails 1'`

**Comment by tschottdorf** [comment]
Note that this issue is relevant only after being rebased on top of #4609.

**Comment by tschottdorf** [comment]
Here's the problem:

```
d0af4e9a-87fd-4083-994f-fac9a66d0fa1  READ  73   1456345378.008884544,2
7ae11e67-1701-436a-94be-1f926cbfd68c  READ  73   1456345378.013854986,0
d0af4e9a-87fd-4083-994f-fac9a66d0fa1  CMTT  74 1456345378.008884544,2
7ae11e67-1701-436a-94be-1f926cbfd68c  CMTT  74 1456345378.013854986,0  (STALE!)
```

Let's call `d0af...` `A` and the other one `B`. A has the lower timestamp and reads first, then B reads - they read the same value. A inserts and commits with a timestamp lower than B. This should not be possible since B scanned "everything" at the higher timestamp. But it is, and so B commits next, leading to the issue.

Presumably there is some inconsistent locking on the read path that allows the following interleaving: B reads, A writes, B updates timestamp cache for its read. We have `(*Replica).readOnlyCmdMu` but basically never use it. I'm relatively confident that that's it.

**Comment by bdarnell** [comment]
`readOnlyCmdMu` was introduced to deal with interactions between reads and splits, not reads and writes. Maybe it needs to be used for writes as well, but that seems like a major limit on concurrency (since it is a replica-wide lock).

Instead of `readOnlyCmdMu`, concurrent access to the timestamp cache is managed by the command queue: If they have overlapping keys, A's write will be queued up behind B's read and won't run until the timestamp cache has been updated.

**Comment by petermattis** [comment]
Could this have been broken by the `RangeGroup` change? Perhaps there is a subtle bug there.

**Comment by tschottdorf** [comment]
This bug is already dead, just doesn't know it yet. I'm also suspecting subtle bug™ in the command queue, just haven't dug in enough yet.

**Comment by tschottdorf** [comment]
Ok, I think I have it. Will send a fix & explanation tomorrow.

**Comment by RaduBerinde** [comment]
Let me enable the test before closing the issue.
